File: 프로그래머스/lv3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    vector<int> answer = vector(enroll.size(), 0);
    map<string, int> index_tree;
    
    for (int i = 0; i < enroll.size(); i++) {
        index_tree[enroll[i]] = i;
    }
    
    for (int i = 0; i < seller.size(); i++) {
        int idx = index_tree[seller[i]];
        int profit = amount[i] * 100;

        while (profit) {
            int commission = profit * 0.1;
            if (referral[idx] == "-") {
                answer[idx] += profit - commission;
                break;
            }
            answer[idx] += profit - commission;
            profit = commission;
            idx = index_tree[referral[idx]];
        }
        
    }
    
    return answer;
}

File: 프로그래머스/lv3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.py
import math

def solution(enroll, referral, seller, amount):
    answer = [0 for _ in range(len(enroll))]
    index_dict = {}
    
    for i in range(len(enroll)):
        index_dict[enroll[i]] = i
    
    for i in range(len(seller)):
        idx = index_dict[seller[i]]
        profit = amount[i] * 10
        answer[idx] += amount[i] * 90
        refer = referral[idx]
        
        while not (refer == '-' or profit < 1):
            idx = index_dict[refer]
            answer[idx] += profit - profit // 10
            refer = referral[idx]
            profit = profit // 10
    
    return answer